File: 3272_find_the_count_of_good_integers.java
class Solution {
    public long countGoodIntegers(int n, int k) {
        int m = (n + 1) / 2, lo = (int)Math.pow(10, m - 1), hi = (int)Math.pow(10, m);
        Set<String> seen = new HashSet<>();
        long res = 0;
        for (int i = lo; i < hi; ++i) {
            String a = i + "", b = new StringBuilder(a).reverse().toString();
            String p = a + b.substring(n % 2);
            if (Long.parseLong(p) % k != 0) continue;
            char[] arr = p.toCharArray();
            Arrays.sort(arr);
            String sig = new String(arr);
            if (!seen.add(sig)) continue;
            int[] f = new int[10];
            for (char c : p.toCharArray()) f[c - '0']++;
            long x = n - f[0], y = 1;
            for (int j = 2; j <= n - 1; ++j) x *= j;
            for (int v : f) for (int j = 2; j <= v; ++j) y *= j;
            res += x / y;
        }
        return res;
    }
}

File: 3272_find_the_count_of_good_integers.cpp
class Solution {
public:
    long long countGoodIntegers(int n, int k) {
        int m = (n + 1) / 2, lo = pow(10, m - 1), hi = pow(10, m);
        unordered_set<string> bag;
        long z = 0;
        for (int i = lo; i < hi; ++i) {
            string s = to_string(i), t = s;
            reverse(t.begin(), t.end());
            s += t.substr(n % 2);
            if (stoll(s) % k) continue;
            string u = s; sort(u.begin(), u.end());
            if (!bag.insert(u).second) continue;
            vector<int> f(10);
            for (char c : s) ++f[c - '0'];
            long p = n - f[0], q = 1;
            for (int j = 2; j < n; ++j) p *= j;
            for (int x : f) for (int j = 2; j <= x; ++j) q *= j;
            z += p / q;
        }
        return z;
    }
};
